summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/FutureQueue.scala

```diff
@@ -0,0 +1,202 @@
+ * via the dequeue method.
+ *
+ * To support batching, inputs can also be TraversableOnce[(S, Future[T])]
+ *
+ * To support additional asynchronous behavior, a state value along with a
+ * Future[TraversableOnce[(S, Future[T])]] can be inserted.  On failure, the
+ * outer state value is returned with the Failure.  On success, the inner
+ * items are queued up to be inserted when they complete.
+ */
+private[summingbird] class FutureQueue[S, T](
+    maxWaitingFutures: MaxWaitingFutures,
+    maxWaitingTime: MaxFutureWaitTime,
+    maxEmitPerExec: MaxEmitPerExecute) {
+  @transient protected lazy val logger: Logger = LoggerFactory.getLogger(getClass)
+
+  // Track the futures that may be outstanding.  We may need to wait on some
+  // of them in case the actual outstanding count gets too high.
+  private lazy val outstandingFutures = Queue.linkedNonBlocking[Future[Unit]]
+  // Track the number of actually outstanding futures
+  private[executor] lazy val numPendingOutstandingFutures = new AtomicInteger(0)
+  // When futures complete, they deposit their results here.
+  private lazy val responses = Queue.linkedNonBlocking[(S, Try[T])]
+
+  private val tryBijection = UtilBijections.twitter2ScalaTry[T]
+
+  /**
+   * Add a Future with state onto the queue.  This can eventually be
+   * dequeued as (S, Try[T]) when the Future completes.
+   *
+   * Returns true if the Future is not yet complete.  When false, the
+   * result can still be dequeued, but this will not count against the
+   * maxWaitingFutures limit.
+   */
+  def add(state: S, fut: Future[T]): Boolean = {
+    val responded =
+      fut.respond { t => responses.put((state, tryBijection(t))) }
+    addOutstandingFuture(responded.unit)
+  }
+
+  /**
+   * Add a collection onto the queue.  Each associated pair can
+   * eventually be dequeued as (S, Try[T]) when the associated
+   * Future completes.
+   *
+   * Returns the number of uncompleted Futures.  All results can
+   * still be dequeued, but only uncompleted Futures will count
+   * against the maxWaitingFutures limit.
+   */
+  def addAll(iter: TraversableOnce[(S, Future[T])]): Int = {
+    val addedSize = iter.count {
+      case (state, fut) => add(state, fut)
+    }
+
+    if (outstandingFutures.size > maxWaitingFutures.get) {
+      /*
+       * This can happen on large key expansion.
+       * May indicate maxWaitingFutures is too low.
+       */
+      logger.debug(
+        "Exceeded maxWaitingFutures({}), put {} futures", maxWaitingFutures.get, addedSize
+      )
+    }
+
+    addedSize
+  }
+
+  /**
+   * Queue a collection Future.  On failure, the state can be retrieved
+   * with the failure.  On success, the results are queued via addAll.
+   *
+   * The Future given here counts against the maxWaitingFutures, so calls
+   * to this method may cause a wait on dequeue.
+   */
+  def addAllFuture(state: S, iterFut: Future[TraversableOnce[(S, Future[T])]]): Unit =
```

Comment:
this seems like a really weird method. I think in the original, with `S = Seq[S1]` then what you actually have is `state` is the union of all states, where each part is separate in `TraversableOnce`.

So, really, I think you have something like, there is a `Monoid[S]` and the sum of all the `TraversableOnce[(S, _)]` is the state.

So, I don't see how this can be general without it.

In the general case, I think we might want something like `Seq[S], Future[Seq[Future[T]]]` so we have a sequence in the same order of states and `T`s. Is that right?